### ISSUE TITLE
Normally-closed end-stops don't need pulldown?

### ISSUE THREAD

**Issue by matthijskooijman** [grbl/grbl#542]
In the FAQ, I read:

> Arduinos have a nice little internal pull-up resistor for normally-open switches that help prevent the pin from frying when the switch is closed. With normally-closed switches, users MUST wire in an external pull-down resistor with their switch. This extra step is usually forgotten and results in a fried pin. 

I'm not sure how this is supposed to work. AFAIU pulldowns / pullups, these prevent the pin from _floating_ when the switch is opened (e.g. to make sure the pin has a predefined logic level when the switch is opened). I can't see how letting a pin float will fry it?

Furthermore, it should be possible to wire in normally-closed switches using the internal pullup as well AFAICS. In this case, the switch normally pulls the pin to ground, so it reads low (and there is a small current, up to 5V/20kohm = 0.25mA running). When the end-stop is activated and the switch is opened, the grounding is broken and the pullup pulls the pin up so it reads a high level. AFAICS the _only_ thing that needs to happen to support normally-closed switches is to invert the pin state when reading it, since now low is idle and high means end-stop activated. The internal pullup can be left on, the switches can connect to GND as normal and no external pulldown is needed.

Can someone point out the flaw in my reasoning, if there is any?

**Comment by matthijskooijman** [comment]
By now, a commit based on my PR was merged by @chamnit (though generalized a bit to also apply to the probe and control pins, nice!). However, I do have some further comments and suggestions wrt limit pins. Some of these might warrant their own issue perhaps, but for now I'll just chunk them together here.

1. The code now contains [this comment][1], about the `DISABLE_*_PIN_PULL_UP`
   macros:

        // The following options disable the internal pull-up resistors,
        // sets the pins to a normal-low operation, and switches much be
        // now connect to Vcc instead of ground. This also flips the
        // meaning of the invert pin Grbl setting, where an inverted
        // setting now means the user should connect a normal-open switch
        // and vice versa.

   On my first reading I was confused, since the meaning of the invert
   setting isn't changed at all - with invert disabled, switches are
   active-low, with invert enabled are active-high.

   However, when I looked more closely, I guess the idea is that with a
   pullup, non-inverted means a NO switch and inverted means
   NC, and with a pulldown, non-inverted means a NC switch and inverted
   means a NO open switch, so the invert setting is flipped in that
   sense.

   Still, this confused me at first, so perhaps this comment needs
   improvement? If it helps, I can propose an umproved version?

   [1]: https://github.com/grbl/grbl/commit/bf37ab7e7be6b6f44a86bf03c223113f46facc7a#diff-98ba56d25d28991fe5fb1f3fe240d0e1R170

2. It seems that the code used for software debouncing, has the meaning
   of the invert pin flipped. Looking at [the code][2], when not
   inverted, it checks if any of the pins reads _high_ and then triggers
   the alarm. So this means limit pins are active-high, though the
   intention is to have the default non-inverted state active-low.

   I can provide a PR to fix this against edge?

   [2]: https://github.com/grbl/grbl/blob/edge/limits.c#L108

3. I previously reported having some stray limit pin triggers. I've
   further investigated and found that these occur randomly, but also
   reliably when switching on or off a nearby industrial vacuum cleaner,
   or when running my mill and the mill bit touches the copper of the
   PCB I'm milling (I'm thinking the copper acts as an antenna for the
   noise generated by the spindle).

   Remembering a comment by @langwadt above on the size of the internal
   pullups, I added external 1K pullups and even a 1uF capacitor to
   ground to form a low-pass filter. This helped reduce noise from the
   vacuum cleaner, but hardly seemed to affect the noise from my spindle
   and PCB. I'm not an expert on this stuff, but I was thinking that
   somehow the noise was picked up after the filter, or the filter was
   too far away from the input pin (I just connected the filter to the
   second limit switch pins, perhaps having the filter on the
   GRBL shield PCB would help).

   Further investigation showed that, even though the PCINT interrupt
   fires, the pin always reads as inactive inside the pin handler. Even
   more, if I write a small sketch that only checks the pin states
   directly in an infinite loop (without an interrupt handler), the pin
   _always_ reads inactive at all. I think this means that the PCINT
   triggering is very sensitive (IIRC it is asynchronous too, so a very
   short pulse can trigger it), but an actual pin read is a better
   indicator of the pin status.

   I've fixed things here by simply reading the limit pin inside the
   PCINT handler - if none of the pins are activated, just return
   without doing anything. This solved all my noise problems, so far
   I've had not any problems with unexpected alarms. Furthermore, the
   alarm always triggered in my testing when pressing the limit switch,
   even when I only very briefly touched it.

   Does it seem reasonable to add this to the main code? If this somehow
   poses a problem for people, we could add a compiletime
   `SENSITIVE_LIMIT_PINS` macro or something to enable the original
   hypersensitive behaviour?

**Comment by chamnit** [comment]
@matthijskooijman : First thanks for bringing up this issue and helping out come up with a solution.

1. You are correct in your assessment. Sure it's a little confusing, but I think this is best suited for a good wiki page explaining how to wire switches correctly for Grbl. I talked with an electrical engineer friend and we discuss the possibilities of the pin configurations and how the potential wiring should be. In general, you'll never need to use the DISABLE_PULLUPS, because of the fact of the internal pull-up are doing a good job. If you do, you are required to wire in your own pull-downs. For most people, this is an extra step they don't need to take.

2. I'm removing that hardware debounce code in v1.0. It's not very useful. If you have a noisy switch, it won't help that much. You should really have good signal conditioning to prevent noise in the first place. 

3. Early on, I was debating whether or not to check the limit state inside the ISR. If the ISR pin change triggers and the switch is bouncing, there is a chance when the ISR checks the pin state, it may read it wrong. To remove any chance of this, I decided not to do it there. This is particularly important, as this ISR is primary for hard limits anyway. Anyhow, if the pin change ISR is hyper sensitive, good signal conditioning is the better solution.

**Comment by matthijskooijman** [comment]
> 1. You are correct in your assessment. Sure it's a little confusing, but I think this is best suited for a good wiki page explaining how to wire switches correctly for Grbl. I talked with an electrical engineer friend and we discuss the possibilities of the pin configurations and how the potential wiring should be. In general, you'll never need to use the DISABLE_PULLUPS, because of the fact of the internal pull-up are doing a good job. If you do, you are required to wire in your own pull-downs. For most people, this is an extra step they don't need to take.

Agreed.

> 2. I'm removing that hardware debounce code in v1.0. It's not very useful. If you have a noisy switch, it won't help that much. You should really have good signal conditioning to prevent noise in the first place.

I take it you mean software debounce here. If it's going to be removed, let's not touch it anymore.

> 3. Early on, I was debating whether or not to check the limit state inside the ISR. If the ISR pin change triggers and the switch is bouncing, there is a chance when the ISR checks the pin state, it may read it wrong. To remove any chance of this, I decided not to do it there. This is particularly important, as this ISR is primary for hard limits anyway. Anyhow, if the pin change ISR is hyper sensitive, good signal conditioning is the better solution.

Hm, I actually bouncing should not cause problems here. Let's explore what could happen (I'm assuming a default active-low setup here). At first glance, a case that seems to cause problems is when the switch is pressed, causing a high-to-low transition, triggering the interrupt. Now, some bouncing happens, and just when the ISR reads the line, the line reads high again, making the ISR think there was no button press after all.

This seems to be a problem, but it's not: if the line reads high, but the button is really pressed, it cannot stay high, so it will cause at least one more high-to-low transition, which triggers the interrupt a second time. Eventually, the interrupt will trigger and read the line as low, triggering the alarm.

Since clearing the interrupt flag happens at the start of the ISR execution, any transitions that happen during the execution of the ISR already cause the flag to be set and the ISR to be triggered again. This means that there is no race condition here (e.g. it's not possible that the high-to-low transition happens after reading a high input, but before the interrupt trigger is enabled again).

Still, I see your point about signal conditioning and I would indeed prefer fixing this properly in the hardware. However, there's two problems:
 - I'm using protoneer's grblshield, which doesn't have any filtering on there. I can add some external components, but the result won't be as nice and robust.
 - I've tried adding an external low-pass filter (a fairly big one too - 1kOhm + 1uF), which didn't actually help. I did just realize that perhaps adding a resistor in series with my switches might also help, since my current filter only has a 1k resistor to VCC. That should slow down low-to-high transitions caused by opening up my NC switch, but I guess it doesn't really help when interference actually drives the switch wire high, running current directly into the cap to charge it.

What kind of signal conditioning are you using on your limit switches?

**Comment by matthijskooijman** [comment]
Thinking a bit more on the series resistor thing, not sure if that would really help - when idle there is 5V/1k = 5mA of current running already, which I think means any interference should first reverse that 5mA before it can actually start charging the capacitor, though I'm not entirely sure...

**Comment by matthijskooijman** [comment]
It seems the series resistor thing might actually be useful, according to [this post](http://www.shapeoko.com/forum/viewtopic.php?f=4&t=1290&start=20#p31054).

Anyway, a bit of Googling shows that there's a lot of people with questions about limit switches, shielding, filtering, etc., so I apologize for the extra noise I'm adding on top of this. For now, let's just say I'll fiddle a bit more with the hardware filtering - feel free to ignore my questions in that area.

However, it would be good if you could have a look at my analysis about fixing this in software - I think that reading the pin in the ISR might actually greatly reduce noise problems for people out of the box, and AFAICS it should not cause alarms to be missed. It probably doesn't solve all noise problems entirely, but if it reduces invalid triggers from a few times per second to a few times per year, or something like that, it's a big win?

**Comment by gerritv** [comment]
A thorough treatise on debouncing switches: http://www.eng.utah.edu/~cs5780/debouncing.pdf 

 

It describes in detail the theory of the 2 resistor and 1 cap method as well as software debounce techniques.

 

Gerrit

**Comment by matthijskooijman** [comment]
@gerritv, thanks for the link, good read. However, this focuses on debouncing switches, which is similar to, but not quite the same as filtering out noise.

As for hardware solutions, I was suspecting a grounding problem. Logic ground and USB ground were already tied together on the Arduino, the grbl shield ties in the stepper ground. I found that the resulting ground showed a 50Hz ripple (10-15Vpp)  compared to the power grid's safety ground (real ground). The stepper power supply I was using is floating, and USB ground came from a laptop. Connecting my laptop's power supply to the reduced the ripple to < 1Vpp, connecting the stepper power supply ground to the safety ground completely removed the ripple. Neither of these fixed my interference problem.

I also tried adding a resistor in series with my switch (as close as possible to the Arduino input pin), which didn't help. I tried a 100 ohm resistor on its own, as well as with a 1k external pullup and 1uF cap to ground. This should really filter out any short pulses, but apparently doesn't.

I still don't really understand why the hardware approaches don't work, but the software approach still works as expected, so let's consider to fix this in software?

**Comment by chamnit** [comment]
@matthijskooijman : Grbl can't have a general solution for everyone and everything. I had to pick something that is guaranteed to stop upon a hard limit. Unless you can guarantee that checking the pin state inside the ISR will not miss a correct reading, then we can discuss this further. But, I would say that you can't, especially with the fact that there are other ISRs, each with their own execution priority levels, operating at the same time. Because your problem is noise, rather than bouncing switches, I would suggest that you try to identify what component is causing the noise and eliminate from your setup.

**Comment by matthijskooijman** [comment]
@chamnit, thanks for your response. I think that my proposed solution will never miss an actual event, see [my analysis above](https://github.com/grbl/grbl/issues/542#issuecomment-73020117). I'm confident it will work, however, I can't guarantee it - if the hardware behaves differently than I think it will, things might behave in an unexpected way.

I understand your concerns, and agree that the default should remain as it is now. However, I'd like to see this added to the code optionally, using a runtime or compiletime option. Would that be acceptable? I suspect that this might be beneficial to others as well - it probably won't be a one-stop-fix for all interference problems, but it adds one more tool to the toolbox.

I also understand that fixing the noise itself would be preferable, but so far all my attempts at this have not been successful so far. Since the software solution seems to fix the problem thoroughly, I'm inclined not to throw more debugging time at it and instead start building things instead :-)

**Comment by chamnit** [comment]
@matthijskooijman : I disagree. I think your code addition suggestion is a patch for a problem specific to your system. (I have several CNCs and none exhibit issues like you have been dealing with.) I'd rather have users correctly use what's already there, than incorrectly trying all of these "additional tools" and blaming Grbl for not working. 

Electrical noise is not a Grbl problem. It's an electrical problem and it should be treated as such.

**Comment by xP373Rx** [comment]
I would suspect the problem somewhere else..  What kind of drivers are you using with the protoner shield?? a4988 or drv8825?

**Comment by matthijskooijman** [comment]
@chamnit, I guess you're right. I'll keep the software change as a local workaround now, but I'll continue looking for the underlying problem.

We're also building a bigger CNC mill in our lab, and two days ago we discovered we were having the same (or at least similar) noise problems on the big machine's end stops. The cable runs are a lot longer and the spindle is a lot bigger, so using a test sketch I counted up to 20,000 unexplained interrupts per second. This allowed me to confirm that the software solution is not an actual solution: Inside the ISR, I still read the wrong pin value a few times per second. I think this means that the difference between the interrupt and reading the pin is not a matter of sensitivity but timing. The interference is strong enough to read as high on both the interrupt and input logic, but the chance of a interference spike happening _exactly_ at the right clock edge is very small. This means that the software approach doesn't actually solve the problem, it just makes it a lot less likely to occur (which is another argument to look further to find a hardware solution).

@xP373Rx, in the small mill, I'm using Polulu modules, I think they are 4983 (the older type).

**Comment by xP373Rx** [comment]
Well fast changing high currents running near logic traces like it is on protoners shield is basicly beging for interferences. Especially if you are  using a high impendance circuit like normal operation pulled up by 2k resistors. I belive its 2k on arduino .. The drivers can be also inducing noise to the 5v circuit  since on protoners shield the arduino have to power the drivers logic side.. 

I am using drv8825 . never had that issue.  The difference on those drivers  is   they power the logic side from the motor power supply.  i also use another shield where the  high current traces are completly separated from the logic side. I dont know how  the GND  is connected to the VMOT GND internally, but they could be some kind of security circuit  between  logic GND pin and VMOT GND pin.

But from mine point of view you, should figure out how to get rid of the noise.. There is no point to deal with it  on the software side.

**Comment by matthijskooijman** [comment]
> I belive its 2k on arduino .. 

It's even 20-50k on the 328 (Arduino Uno)

I did try adding 1k pullup and 1uF cap, without any luck. I suspect that sticking them onto the pin headers on the shield might leave them too far from the input pin and too open. Soldering them onto the shield would probably improve things, but then I'd have to redesign the shield.

As for leakage from the stepper motor supply, I think I was able to trigger the problem even with the steppers disabled, perhaps even with the stepper supply powered off, generating interference with my spindle. I'd have to doublecheck this, though.

I guess I should also improve the cabling. I'm now using a single twisted pair, probably using some shielded cable would improve things. AFAIU I should make sure to connect the shield to (Arduino) ground on the Arduino side, but leave it disconnected on the other end, to prevent the return current from using the shield? I don't have a proper chassis ground to connect the shield to, my machine has a wooden casing :-)

**Comment by xP373Rx** [comment]
well i did the worst i coud do   and put switches  in normal mode shorted to the ground and reprogrammed grbl to  having pullup on when pin is inverted.. it works fine and you have less then 1 ohm impendance.. but to be safe you could add  resistors  to prevent fraying  the pins 470Ω or 1k to prevent fraying when the pin is accidently switched to output mode.

**Comment by gerritv** [comment]
@matthijskooijman 
It takes 2 resistors and a cap to make a proper debounce/filter circuit. the cap is likely to be at 0.1uf. Page 12 of http://www.eng.utah.edu/~cs5780/debouncing.pdf gives the details and math.
If your motor has brushes, those need to be filtered with a cap each as well. If you have relays, use a diode and caps to prevent spikes.

Gerrit

**Comment by poelstra** [comment]
You could be having a ground-loop issue.

I too had an issue with occassionally triggering hard limits and the reset/abort pin seeming to be triggered without a cause. I never had that when I was running LinuxCNC, but after adding debug info to Grbl it really was pointing to an electrical problem.

When I switched to Grbl, I also switched to a new laptop for controlling it, and this new laptop has a three-pronged plug on its power-supply, effectively closing a ground loop. Unplugging the laptop's power also fixed the electrical problems. Probably have to find a better fix some day, but for now I'll simply plug my laptop's power-supply into an unearthed socket...

**Comment by matthijskooijman** [comment]
@poelstra, I considered that option, but I'm also having problems when my laptop is not plugged in and running on battery. I tried various variations in the grounding (grounding through my laptop power supply, grounding through the stepper supply, both or neither), with no effect on my problem...

@xP373Rx has contacted me in private with a few more suggestions of things to try and measure. I'll report back when I did those, but I haven't been near the machine for a while now :-)

In any case, thanks all for the suggestions!
